Isosurfacing_3/include/CGAL/Isosurfacing_3/internal/marching_cubes_functors.h

```diff
@@ -0,0 +1,297 @@
+
+    // if(e_tris_list[t_index] == 0x7f)
+    if(Cube_table::triangle_cases[t_index] == -1)
+      break;
+
+    // @todo move more of this stuff into the table
+    const int eg0 = Cube_table::triangle_cases[t_index + 0];
+    const int eg1 = Cube_table::triangle_cases[t_index + 1];
+    const int eg2 = Cube_table::triangle_cases[t_index + 2];
+
+    // insert new triangle in list
+    #ifdef CGAL_LINKED_WITH_TBB
+      auto& tris = triangles.local();
+    #else
+      auto& tris = triangles;
+    #endif
+    tris.push_back({vertices[eg0], vertices[eg1], vertices[eg2]});
+  }
+}
+
+template <typename TriangleRange,
+          typename PointRange,
+          typename PolygonRange>
+void triangles_to_polygon_soup(const TriangleRange& triangles,
+                               PointRange& points,
+                               PolygonRange& polygons)
+{
+  #ifdef CGAL_LINKED_WITH_TBB
+    for(const auto& triangle_list : triangles)
+    {
+  #else
+      const auto& triangle_list = triangles;
+  #endif
+
+      for(const auto& triangle : triangle_list)
+      {
+        const std::size_t id = points.size();
+
+        points.push_back(triangle[0]);
+        points.push_back(triangle[1]);
+        points.push_back(triangle[2]);
+
+        // simply use increasing indices
+        polygons.push_back({id + 2, id + 1, id + 0});
+      }
+
+  #ifdef CGAL_LINKED_WITH_TBB
+    }
+  #endif
+}
+
+// Marching Cubes implemented as a functor that runs on every cell of the grid
+template <typename Domain_>
+class Marching_cubes_3
+{
+public:
+  using Domain = Domain_;
+
+  using Geom_traits = typename Domain::Geom_traits;
+  using FT = typename Geom_traits::FT;
+  using Point_3 = typename Geom_traits::Point_3;
+
+  using Cell_descriptor = typename Domain::Cell_descriptor;
+
+#ifdef CGAL_LINKED_WITH_TBB
+  using Triangles = tbb::enumerable_thread_specific<std::vector<std::array<Point_3, 3>>>;
+#else
+  using Triangles = std::vector<std::array<Point_3, 3> >;
+#endif
+
+private:
+  const Domain& m_domain;
+  const FT m_isovalue;
+
+  Triangles m_triangles;
+
+public:
+  // creates a Marching Cubes functor for a domain and isovalue
+  Marching_cubes_3(const Domain& domain,
+                   const FT isovalue)
+    : m_domain(domain),
+      m_isovalue(isovalue)
+  { }
+
+  // gets the created triangle list
+  Triangles& triangles()
+  {
+    return m_triangles;
+  }
+
+public:
+  // computes one cell
+  void operator()(const Cell_descriptor& cell)
+  {
+    // @todo: maybe better checks if the domain can be processed?
+    CGAL_precondition(m_domain.cell_vertices(cell).size() == 8);
+    CGAL_precondition(m_domain.cell_edges(cell).size() == 12);
+
+    std::array<FT, 8> values;
+    std::array<Point_3, 8> corners;
+    const int i_case = get_cell_corners(m_domain, cell, m_isovalue, corners, values);
+
+    // skip empty cells
+    const int all_bits_set = (1 << (8 + 1)) - 1; // last 8 bits are 1
+    if(i_case == 0 || i_case == all_bits_set)
+      return;
```

Comment:
Because we have this potential early exit, I would suggest that `get_cell_corners()` gets split in two functions.  One for the isovalues, and another one for the points, as the latter are not needed that often after all.